File: src/main/java/net/guizhanss/fastmachines/FastMachines.java
package net.guizhanss.fastmachines;

import java.util.logging.Level;

import javax.annotation.Nonnull;

import com.google.common.base.Preconditions;

import io.github.thebusybiscuit.slimefun4.libraries.dough.updater.BlobBuildUpdater;

import net.guizhanss.fastmachines.core.Registry;
import net.guizhanss.fastmachines.core.services.ConfigurationService;
import net.guizhanss.fastmachines.core.services.IntegrationService;
import net.guizhanss.fastmachines.core.services.ListenerService;
import net.guizhanss.fastmachines.core.services.LocalizationService;
import net.guizhanss.fastmachines.setup.Items;
import net.guizhanss.fastmachines.setup.Researches;
import net.guizhanss.guizhanlib.slimefun.addon.AbstractAddon;
import net.guizhanss.guizhanlib.updater.GuizhanBuildsUpdater;

import org.bstats.bukkit.Metrics;

public final class FastMachines extends AbstractAddon {

    public static final String DEFAULT_LANG = "en-US";

    private Registry registry;
    private ConfigurationService configService;
    private LocalizationService localization;
    private IntegrationService integrationService;
    private boolean debugEnabled = false;

    public FastMachines() {
        super("ybw0014", "FastMachines", "master", "auto-update");
    }

    @Nonnull
    public static Registry getRegistry() {
        return inst().registry;
    }

    @Nonnull
    public static ConfigurationService getConfigService() {
        return inst().configService;
    }

    @Nonnull
    public static LocalizationService getLocalization() {
        return inst().localization;
    }

    @Nonnull
    public static IntegrationService getIntegrationService() {
        return inst().integrationService;
    }

    public static void debug(@Nonnull String message, @Nonnull Object... args) {
        Preconditions.checkNotNull(message, "message cannot be null");

        if (inst().debugEnabled) {
            inst().getLogger().log(Level.INFO, "[DEBUG] " + message, args);
        }
    }

    @Nonnull
    private static FastMachines inst() {
        return getInstance();
    }

    @Override
    public void enable() {
        log(Level.INFO, "====================");
        log(Level.INFO, "     FastMachines   ");
        log(Level.INFO, "     by ybw0014     ");
        log(Level.INFO, "====================");

        // check sf version
        if (!checkSlimefunVersion()) return;

        // registry
        registry = new Registry();

        // config
        configService = new ConfigurationService(this);

        // debug
        debugEnabled = configService.isDebugEnabled();

        // localization
        log(Level.INFO, "Loading language...");
        String lang = configService.getLang();
        localization = new LocalizationService(this, getFile());
        try {
            localization.addLanguage(lang);
        } catch (Exception e) {
            log(Level.SEVERE, "An error has occurred while loading language " + lang);
        }
        if (!lang.equals(DEFAULT_LANG)) {
            localization.addLanguage(DEFAULT_LANG);
        }
        localization.setIdPrefix("FM_");
        log(Level.INFO, localization.getString("console.loaded-language"), lang);

        // items
        log(Level.INFO, localization.getString("console.loading-items"));
        Items.setup(this);

        // integrations
        integrationService = new IntegrationService(this);

        // researches
        if (configService.isEnableResearches()) {
            log(Level.INFO, localization.getString("console.loading-researches"));
            Researches.setup();
            Researches.register();
        }

        new ListenerService(this);

        setupMetrics();
    }

    @Override
    public void disable() {
        // nothing to do here for now
    }

    private void setupMetrics() {
        new Metrics(this, 20046);
    }

    @Override
    protected void autoUpdate() {
        if (getPluginVersion().startsWith("Dev")) {
            new BlobBuildUpdater(this, getFile(), getGithubRepo()).start();
        } else if (getPluginVersion().startsWith("Build")) {
            new GuizhanBuildsUpdater(this, getFile(), getGithubUser(), getGithubRepo(), getGithubBranch()).start();
        }
    }

    private boolean checkSlimefunVersion() {
        String sfVersion = getServer().getPluginManager().getPlugin("Slimefun").getDescription().getVersion();
        if (sfVersion.startsWith("DEV - 1104")) {
            for (int i = 0; i < 100; i++) {
                log(Level.SEVERE, "You are using a damn ancient version of Slimefun, update Slimefun first!");
                log(Level.SEVERE, "Download Slimefun from here: https://blob.build/project/Slimefun4/Dev");
                log(Level.SEVERE, "Also join the Slimefun discord so you can get the announcements for new versions and will not be a primitive any more.");
                log(Level.SEVERE, "https://discord.gg/slimefun");
            }
            getServer().getPluginManager().disablePlugin(this);
            return false;
        }
        return true;
    }
}

File: src/main/kotlin/net/guizhanss/fastmachines/FastMachines.kt
package net.guizhanss.fastmachines

import io.github.thebusybiscuit.slimefun4.libraries.dough.updater.BlobBuildUpdater
import net.byteflux.libby.Library
import net.guizhanss.fastmachines.core.FMRegistry
import net.guizhanss.fastmachines.core.services.ConfigService
import net.guizhanss.fastmachines.core.services.IntegrationService
import net.guizhanss.fastmachines.core.services.LocalizationService
import net.guizhanss.fastmachines.implementation.groups.FMItemGroups
import net.guizhanss.fastmachines.implementation.items.FMItems
import net.guizhanss.fastmachines.implementation.listeners.CauldronListener
import net.guizhanss.fastmachines.implementation.listeners.GravityListener
import net.guizhanss.fastmachines.implementation.listeners.HopperListener
import net.guizhanss.fastmachines.implementation.listeners.SlimefunRegistryListener
import net.guizhanss.fastmachines.implementation.setup.ResearchSetup
import net.guizhanss.fastmachines.implementation.tasks.FastMachineTickingTask
import net.guizhanss.guizhanlib.libraries.BukkitLibraryManager
import net.guizhanss.guizhanlib.slimefun.addon.AbstractAddon
import net.guizhanss.guizhanlib.updater.GuizhanBuildsUpdater
import org.bstats.bukkit.Metrics
import org.bukkit.Bukkit
import org.bukkit.plugin.Plugin
import java.io.File
import java.util.logging.Level

class FastMachines : AbstractAddon(
    GITHUB_USER, GITHUB_REPO, GITHUB_BRANCH, AUTO_UPDATE_KEY
) {

    override fun load() {
        // check if there is central repo prop defined
        val centralRepo = System.getProperty("centralRepository") ?: "https://repo1.maven.org/maven2/"

        logger.info("Loading libraries, please wait...")
        logger.info("If you stuck here for a long time, try to specify a mirror repository.")
        logger.info("Add -DcentralRepository=<url> to the JVM arguments.")

        // download libs
        val manager = BukkitLibraryManager(this)
        manager.addRepository(centralRepo)
        manager.loadLibrary(
            Library.builder().groupId("org.jetbrains.kotlin").artifactId("kotlin-stdlib").version("2.1.10").build()
        )
        manager.loadLibrary(
            Library.builder().groupId("org.jetbrains.kotlin").artifactId("kotlin-reflect").version("2.1.10").build()
        )

        logger.info("Loaded all required libraries.")
    }

    override fun enable() {
        instance = this

        FMRegistry

        // config
        configService = ConfigService(this)
        debug("Debug mode is enabled.")

        // localization
        log(Level.INFO, "Loading language...")
        val lang = configService.lang
        localization = LocalizationService(this, file)
        localization.idPrefix = "FM_"
        localization.addLanguage(lang)
        if (lang != DEFAULT_LANG) {
            localization.addLanguage(DEFAULT_LANG)
        }
        log(Level.INFO, "Loaded language {0}.", lang)

        // integrations
        integrationService = IntegrationService(this)

        // item groups setup
        FMItemGroups

        // item setup
        FMItems

        // researches setup
        if (configService.enableResearches) {
            ResearchSetup
        }

        // listeners & tasks
        setupListeners()
        setupTasks()

        // Metrics setup
        setupMetrics()
    }

    override fun disable() {
        Bukkit.getScheduler().cancelTasks(this)
    }

    override fun autoUpdate() {
        if (pluginVersion.startsWith("Dev")) {
            BlobBuildUpdater(this, file, githubRepo).start()
        } else if (pluginVersion.startsWith("Build")) {
            try {
                // use updater in lib plugin
                val clazz = Class.forName("net.guizhanss.minecraft.guizhanlib.updater.GuizhanUpdater")
                val updaterStart = clazz.getDeclaredMethod(
                    "start",
                    Plugin::class.java,
                    File::class.java,
                    String::class.java,
                    String::class.java,
                    String::class.java
                )
                updaterStart.invoke(null, this, file, githubUser, githubRepo, githubBranch)
            } catch (ignored: Exception) {
                // use updater in lib
                GuizhanBuildsUpdater.start(this, file, githubUser, githubRepo, githubBranch)
            }
        }
    }

    private fun setupListeners() {
        CauldronListener(this)
        GravityListener(this)
        HopperListener(this)
        SlimefunRegistryListener(this)
    }

    private fun setupTasks() {
        FastMachineTickingTask()
    }

    private fun setupMetrics() {
        Metrics(this, 20046)
    }

    companion object {

        private const val GITHUB_USER = "ybw0014"
        private const val GITHUB_REPO = "FastMachines"
        private const val GITHUB_BRANCH = "master"
        private const val AUTO_UPDATE_KEY = "auto-update"
        const val DEFAULT_LANG = "en-US"

        lateinit var instance: FastMachines
            private set
        lateinit var configService: ConfigService
            private set
        lateinit var localization: LocalizationService
            private set
        lateinit var integrationService: IntegrationService
            private set

        fun scheduler() = getScheduler()

        val slimefunTickCount: Int
            get() = getSlimefunTickCount()

        fun log(level: Level, message: String) {
            instance.logger.log(level, message)
        }

        fun log(level: Level, ex: Throwable, message: String) {
            instance.logger.log(level, ex) { message }
        }

        fun debug(message: String) {
            if (!Companion::configService.isInitialized || !configService.debug) return
            log(Level.INFO, "[DEBUG] $message")
        }
    }
}
